### PR TITLE
feat: Pytorch2 necessary changes

### PR DESCRIPTION
**Pull request by MikhailKardash**
## Description

A few changes preceding the advent of `pytorch==2.0.1` images:
1. A versioning conditional on our PyTorch DataLoader
2. Deprecated `attrdict` replaced with `attrdict3` in harness. This is necessary for `python>=3.9`. Either library is fine for `python==3.8`.

I've also thrown in a small update to `update-bumpenvs-yaml.py` to remove checking for `tf1` images following an update to our environments repo.

## Test Plan

All PyTorch and DeepSpeed related tests should pass in CI.

## Commentary (optional)

Upgrading our task environment to `pytorch==2.0.1` is not as impossible as it seems, but we will have to take a look at `apex.amp` and our old, version pinned examples, as those are the key blockers of this from an OSS standpoint.

## Checklist

- [X] Changes have been manually QA'd
- [X] User-facing API changes need the "User-facing API Change" label.
- [X] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [X] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: harness/determined/pytorch/_data.py [hunk 1]
```diff
@@ -245,13 +248,18 @@ def get_data_loader(
             batch_sampler, repeat=repeat, skip=skip, num_replicas=num_replicas, rank=rank
         )
 
-        # Try to no break any torch version as old as v1.0.
+        # Try to not break any torch version as old as v1.0.
```
Comment (ioga) [inline harness/determined/pytorch/_data.py:251]: since we stopped supporting torch 1.7... remove lines 253–257?

**Review by ioga — APPROVED** [review]

File: harness/determined/pytorch/_data.py (hunk 1 above)
Comment (MikhailKardash) [reply harness/determined/pytorch/_data.py]: Done, I removed anything that references version<1.10 in this file. :)

Comment (rb-determined-ai) [reply harness/determined/pytorch/_data.py]: Wait, I know this is after-merge, but when did we stop supporting torch 1.7?

There's a difference between "our environments support 1.7" and "our APIs support 1.7".  I think, if torch is api-compatible to code written against 1.0, we should still be supporting those versions, at least at the API layer.

Comment (ioga) [reply harness/determined/pytorch/_data.py]: we had a note in 0.22.0 saying that we’ve stopped supporting tf 1 and torch 1.7: https://docs.determined.ai/latest/release-notes.html#version-0-22-0

the note does not make distinction between environments and APIs, TF1 APIs will stop being supported quite explicitly. when we discussed it, I don't recall anyone coming in defence of torch 1.7. if anything, people are asking for newer pytorch versions all the time. so I think the users have been notified appropriately in this case.

Comment (rb-determined-ai) [reply harness/determined/pytorch/_data.py]: Ok, I think the note is wrong in its approach then.

Not supporting those versions in our environments? Sure.

Not supporting tf1 in the keras code we own?  Sure.  That support was expensive, and tf1 vs tf2 was a huge breaking change.

Not supporting Pytorch 1.0 in the pytorch code we own?  I don't agree with that.  Pytorch continues to be stable back to 1.0, and so I think the (very small) vendored pytorch code we have should also have that kind of stability.

It costs nothing to just leave it alone, and it might help somebody someday.

It might never help anybody, too.

I don't see where we draw the line on "how recent of pytorch do we support" with the system you are suggesting.  I don't want our oldest environments to be the oldest thing we support.  I'd like us to update environments much more frequently, but to continue to work with old pytorch just like pytorch itself does.  I'd like more customers with those situations to feel safe owning their own custom images and expecting them to work, because that decreases our environments burden.

But tying our pytorch api support to the pytorch environments we ship... that makes changing environments more expensive and more impactful to customers, and I think that's a bad idea.
